### ISSUE TITLE
vim-gitgutter highlights incorrect lines when files are transformed via filter

### ISSUE THREAD

**Issue by systemmonkey42** [airblade/vim-gitgutter#796]
Hi All,

In the same vein as #435, I am using a filter to convert files as they enter/leave GIT.   This results in git-gutter highlighting almost everything in the file as changed.

The conversion I am doing results in the file in git looking quite different, however git-diff is unaffected.

I have edited the file gitgutter/diff/diff.vim to add the option `--textconv` to the 'git show' command, thus
ensuring all text conversion which diff would normally see is performed on the file before being returned for
vim-gitgutter to process.

Is there a downside to using `--textconv` in this way?   I understand it might be a new option compared to what was available when answering #435.  Would making it the default behaviour be a reasonable choice?

Thanks

**Comment by airblade** [comment]
I've never used filters so I don't know how they work in any great detail.

When you use `--textconv` are the line numbers in the git-diff output correct for the vim buffer?

>  For this reason, only git diff and the git log family of commands (i.e., log, whatchanged, show) will perform text conversion

– [Performing text diffs of binary files](https://git-scm.com/docs/gitattributes#_performing_text_diffs_of_binary_files)

It sounds like it might not be possible to stage / undo hunks.

**Comment by systemmonkey42** [comment]
> When you use --textconv are the line numbers in the git-diff output correct for the vim buffer?

Yes.   The logic is that GIT performs a text conversion when you first check out the file, and performs that conversion in the background when calling git-diff.  The output of `git show --textconv` is exactly what you get when you checkout the file.

> It sounds like it might not be possible to stage / undo hunks.

This.  A "patch" prepared from the "smudged" copies cannot be applied.   Git will "clean" the working file and diff that with the existing blob when staging hunks.  Even `git add -p` doesn't work as expected.

I'm just happy that the file changes highlight correctly and it allows me to use `]c` and `[c` to navigate changes.

**Comment by airblade** [comment]
That's all good to know.

Do you have steps for setting up a simple filter so I can experiment please?

**Comment by systemmonkey42** [comment]
I put together a contrived example using gitattributes and filters.

https://github.com/systemmonkey42/attribute-test

This uses two filters

bash.1 - filtered via "base64", hence the contents of GIT are completely different to what you see after checkout

bash.2 - filtered via "rev".   The contents in git are the same length, same number of lines, same number of characters, each line of text is simply reversed.

As i said, contrived.  But it should help demonstrate what is going on.

Hint:  use `export GIT_TRACE=1` to confirm git is actually running the filters on each command.

Please refer to the README in the repository for instructions on enabling the filters.   A feature of GIT is that the configuration changes required for filters to work cannot be part of the repository itself.   For security reasons of course.

**Comment by airblade** [comment]
Thanks for this, much appreciated.  I'll take a look as soon as I get some time.

**Comment by airblade** [comment]
*Almost two years goes by...*

Your demo repo is very helpful.  I'm not sure I have set it up correctly – please see https://github.com/systemmonkey42/attribute-test/issues/1.

Anyway, it looks like adding `--textconv` to `git-show` was the only change you needed for this:

https://github.com/airblade/vim-gitgutter/compare/master...systemmonkey42:vim-gitgutter:master

Have you found any downsides over the past couple of years?

I was thinking it would be nice to detect when text conversion filters are active on a file so that gitgutter can disable the hunk staging.  The only way I can see to determine this is to parse `.gitattributes` and compare to the filename in question, which sounds a little fiddly.

**Comment by systemmonkey42** [comment]
> Your demo repo is very helpful. I'm not sure I have set it up correctly – please see 

I believe you set it up correctly.   Git can behave a little strangely if the attributes are set up
after checkout, because the index/stage/worktree get out of sync.

My reply shows a few workarounds, my favourite being "git update-index" because it almost always straightens out the mess.

> Have you found any downsides over the past couple of years?

Apart from an inability to stage patches, it has all worked perfectly.

I agree detecting if a "filter" is in play, and showing an error for "stage" and "revert" commands will be the best option.  Everything else works perfectly.

Also, git has a core function for checking attributes. You probably want the following:
```
$ git check-attr filter -- bash.2
bash.2: filter: rev
```
vs

A file with no filter attribute:
```
$ git check-attr filter -- install.sh
install.sh: filter: unspecified
```

You can also use `git check-attr -a` and just filter for the "filter" keyword, though technically any attribute (such as `eol=crlf`, or `-text`, etc) might have an impact on staging hunks...

**Comment by airblade** [comment]
> You can also use `git check-attr -a` and just filter for the "filter" keyword, though technically any attribute (such as `eol=crlf`, or `-text`, etc) might have an impact on staging hunks...

I think I'll look for "filter", ignore the rest, and hope for the best.  GitGutter already tries to [deal with line endings](https://github.com/airblade/vim-gitgutter/blob/11d6e13194bbbc27dbfefdca4d71099804e70b48/autoload/gitgutter/diff.vim#L38-L62) and it seems to work, in that nobody has opened an issue about it.

**Comment by airblade** [comment]
Ah, it works when there is only one hunk.  If there are, say, two, the binary diff still shows a single change.  Which makes sense.  So we can only stage the whole file, not part of it, as we originally thought.

**Comment by airblade** [comment]
> Thinking about this, a "File uses filter. Stage entire file?" prompt allowing me to effectively `git add <file>` might be a nice to have.

It does this now.  Please try it out and let me know if it solves your problem.

**Comment by systemmonkey42** [comment]
Hi, Thanks for that.

I get the message `File uses clean/smudge filter. Stage entire file (y/n)?`

However after I choose "y", I get `Patch does not apply` 

I tried to set g:gitgutter_log=1, but that results in 
```
Error detected while processing function <lambda>2[1]..gitgutter#process_buffer[33]..function <lambda>2[1]..gitgutter#process_buffer[23]..script /root/.vim/pack/bundles/opt/vim-gitgutter/autoload/gitgutter/diff.v
im[13]..function <SNR>271_git_supports_command_line_config_override[1]..gitgutter#utility#system[1]..gitgutter#debug#log:
line    8:
E930: Cannot use :redir inside execute()
```

**Comment by airblade** [comment]
I have fixed the logging in 621b9e7.

Thanks for diagnosing the path problem.  I hadn't noticed that when testing with your demo repo because the files are in the root directory.

GitGutter always changes to a file's directory when running commands to ensure that git figures out the correct repo.  I have updated the code (7d425ff) to run `cd path/to && git add file` instead of `cd path/to && git add path/to/file`.
